api/internal/target/kusttarget_configplugin.go

```diff
@@ -108,339 +110,332 @@ func (kt *KustTarget) configureBuiltinTransformers(
 	return result, nil
 }
 
-type gFactory func() resmap.GeneratorPlugin
+// GetGeneratorConfigurators returns configurators for built-in generators with corresponding kustomization fields.
+// The configurators only operate on non-deprecated kustomization fields.
+func GetGeneratorConfigurators() map[builtinhelpers.BuiltinPluginType]func(*types.Kustomization, *resmap.PluginHelpers) (
+	[]resmap.Generator, error) {
 
-var generatorConfigurators = map[builtinhelpers.BuiltinPluginType]func(
-	kt *KustTarget,
-	bpt builtinhelpers.BuiltinPluginType,
-	factory gFactory) (result []resmap.Generator, err error){
-	builtinhelpers.SecretGenerator: func(kt *KustTarget, bpt builtinhelpers.BuiltinPluginType, f gFactory) (
-		result []resmap.Generator, err error) {
-		var c struct {
-			types.SecretArgs
-		}
-		for _, args := range kt.kustomization.SecretGenerator {
-			c.SecretArgs = args
-			c.SecretArgs.Options = types.MergeGlobalOptionsIntoLocal(
-				c.SecretArgs.Options, kt.kustomization.GeneratorOptions)
-			p := f()
-			err := kt.configureBuiltinPlugin(p, c, bpt)
-			if err != nil {
+	configurator := func(plugin builtinhelpers.BuiltinPluginType, h *resmap.PluginHelpers, inputs [][]byte) (
+		[]resmap.Generator, error) {
+		generators := make([]resmap.Generator, len(inputs))
+		for i, input := range inputs {
+			configurable := builtinhelpers.GeneratorFactories[plugin]()
+			if err := configurable.Config(h, input); err != nil {
 				return nil, err
 			}
-			result = append(result, p)
+			generators[i] = configurable
 		}
-		return
-	},
+		return generators, nil
+	}
 
-	builtinhelpers.ConfigMapGenerator: func(kt *KustTarget, bpt builtinhelpers.BuiltinPluginType, f gFactory) (
-		result []resmap.Generator, err error) {
-		var c struct {
-			types.ConfigMapArgs
-		}
-		for _, args := range kt.kustomization.ConfigMapGenerator {
-			c.ConfigMapArgs = args
-			c.ConfigMapArgs.Options = types.MergeGlobalOptionsIntoLocal(
-				c.ConfigMapArgs.Options, kt.kustomization.GeneratorOptions)
-			p := f()
-			err := kt.configureBuiltinPlugin(p, c, bpt)
-			if err != nil {
-				return nil, err
+	return map[builtinhelpers.BuiltinPluginType]func(*types.Kustomization, *resmap.PluginHelpers) (
```

Comment:
It's not clear to me why all this is necessary. It seems a bit overly complicated to me. What I was imagining was that the localizer could:

- process each kustomization field, one by one
- determine if the field is powered by a transformer or a generator. 
- If so, marshal the kustomization field into the generator struct, and call `MapFiles`. 

None of this implies touching `kusttarget` or changing anything about the plugins. Am I missing something?

Reply:
Just a high-level response: 
* I changed the `generatorConfigurators` because 
  * I want the plain `ConfigMapGenerator`, `SecretGenerator` without the `GeneratorOptions`, which means I need to pass in a `kustomization` instead of `kusttarget` to the configurators. Alternatively, I could modify the returned plugin, or write my own configurator for these 2 cases, but these solutions seem more convoluted.
  * `lint` complains about the global variable, so I wrapped it in a function
* I need `YAML()` because I need to be able to serialize the plugin to write it both
  * back to the kustomization
  * to the files referenced under kustomization fields `generators` and `transformers`
* I modified `Generate()`, `Transform()` because they were changing struct fields that would've impacted the output of `YAML()`. Also, in our discussion, @natasha41575, I remember you recommending `MapFiles()` not performing file system tasks like loading file contents. As a result, in order to keep the plugin struct states consistent, I moved loader logic into the `Transform()` methods.

Reply:
> Also, in our discussion, @natasha41575, I remember you recommending MapFiles() not performing file system tasks like loading file contents. As a result, in order to keep the plugin struct states consistent, I moved loader logic into the Transform() methods.

I've read this a few times and I'm not sure I understand. `MapFiles` should be completely independent from `Transform`. `MapFiles` should only be used by `localize`, while `Transform` should only be used by `build`.

Reply:
Yes, but `MapFiles` and `Configure` are called together, and most of this logic used to be in `Configure`. I moved it into `Transform`. 

On another note, `MapFiles()` and `Transform()` aren't currently called together, but they're on the same struct and so theoretically can be called in any order. Once we add `MapFiles()`, we don't know if/how future developers will use it. We can check during code reviews that no new code is using them together, but that's not very safe. There's no hard inhibitor in place (unless we add it) preventing their call order, and so IMO they should work with each other.

Reply:
> Yes, but MapFiles and Configure are called together, and most of this logic used to be in Configure. I moved it into Transform.

IMO the logic should stay in `Configure`. `localize` can have its own code for what it needs, and common parts can move into some helper function.

> On another note, MapFiles() and Transform() aren't currently called together, but they're on the same struct and so theoretically can be called in any order. Once we add MapFiles(), we don't know if/how future developers will use it. We can check during code reviews that no new code is using them together, but that's not very safe. There's no hard inhibitor in place (unless we add it) preventing their call order, and so IMO they should work with each other.

We don't need to worry about this now. If a future developer wants to use `MapFiles` and `Transform` together, they can make the changes they need themselves.

Reply:
These concerns should be resolved by the new implementation.